### ISSUE TITLE
Create a graph from an edge list

### ISSUE THREAD

**Issue by sbromberger** [JuliaGraphs/LightGraphs.jl#875]
Currently, we do not have a constructor for `SimpleGraph`/`SimpleDiGraph` that can build a graph from an edgelist. This should be a fairly straightforward constructor:

From #755, which was the genesis of this idea:

> I'd like to revisit the premise of this function to see whether we can get there without the churn. It seems to me that we need something analogous to NetworkX's add_edges_from method. If this is the case, then I would propose creating a constructor that most efficiently creates a (di)graph from an edge list. We could then use blkdiag or union with this graph and an existing graph if you wanted the result to be "add an edge list to an existing graph".

cc @mschauer @jpfairbanks

**Comment by mschauer** [comment]
Note that #755 already contains an implementation for building a SimpleGraph/SimpleDiGraph from an iterator of edges of possibly unknown length which can easily be adapted to create a new graph after sorting out where to throw errors or not.

**Comment by simonschoelly** [comment]
After looking at the source code, I think I could try to implement this. There are some issues with the type system though:

Right now, we the following constructors for SimpleGraph:
```
SimpleGraph()
SimpleGraph(n::T) where T<:Integer
SimpleGraph(::Type{T}) where T<:Integer

SimpleGraph{T}(g::SimpleGraph) where T<:Integer
SimpleGraph(g::SimpleDiGraph)

SimpleGraph(adjmx::AbstractMatrix) 
SimpleGraph{T}(adjmx::AbstractMatrix) where T<:Integer
```
Ideally I would be able to add a constructor like `SimpleGraph(Collection{Pair})`, (where Collection can be an Array, Vertex, Set, Dict, Generator, ....) but Julia's type system is not strong enough to express this. I though of 4 different options that all have their drawbacks:

1.) Use a constructor of the form `SimpleGraph(edge_list)` that matches whenever no other constructor matches. This has the drawback, that in some cases it could lead to a mix-up with `SimpleGraph(adjmx::AbstractMatrix) `.  Furthermore, in the future someone might want to add the constructor `SimpleGraph(adj_list)` that creates a graph from an adjacency list.

2.) Write specialized constructors for all different kind of collections, such as  `SimpleGraph(edge_list::Set{Pair{Integer, Integer}} )`  and `SimpleGraph(edge_list::Dict{Integer, Integer}} )` . This might lead to a lot of duplicated code.

3.) Don't use the constructor but instead create a function `createSimpleGraph_from_edgelist(...)` (would need a better name).

4.) Use kwargs in the constructor: `SimpleGraph(; edge_list=[ ... ])` .  I don't know if this good style though.

Any thoughts?

**Comment by sbromberger** [comment]
My preference would be to create `SimpleGraph(edge_list::Vector{Edge})` and possibly let folks do `Edge.(array)` if they have their edges in some other format (like tuples):

```
julia> a = [(1,2), (3,4)]
2-element Array{Tuple{Int64,Int64},1}:
 (1, 2)
 (3, 4)

julia> Edge.(a)
2-element Array{LightGraphs.SimpleGraphs.SimpleEdge{Int64},1}:
 Edge 1 => 2
 Edge 3 => 4
```

Comments?

**Comment by mschauer** [comment]
It's like to keep room open for the possibility to have an lazy iterator giving the egdes in order not to construct graphs two times in memory. A way would be to query the `eltype`-trait and dispatch on that:
```
SimpleGraph(e) = _SimpleGraph(e, IteratorEltype(typeof(e)))
 _SimpleGraph(e, ::Type{Edge}) = ... # Edges
 _SimpleGraph(e::Matrix, ::Type{Int}) = ... # Adjacency matrix
```

**Comment by sbromberger** [comment]
I like the idea of using an iterator as well, but I think it might be easier just to create a separate constructor for it. (The alternative is that we can make the iterator-based constructor the "primary" one and have the `Vector{Edge}` constructor just call it with an iterator over the vector.)

**Comment by simonschoelly** [comment]
Well, I've written now a function `SimpleGraph(edge_list::Vector{SimpleGraphEdge{T}}) where T<:Integer` that
constructs a graph from a vector of edges and a function  `SimpleGraphFromIterator(iter)` that constructs a graph from an arbitrary iterator over edges. Then I did some benchmarks:

```
function run_benchmarks(n::Int64)
    iter = (Edge(i, j) for i in 1:n for j in (i+1):n )

    @time g1 = CompleteGraph(n)
    @time g2 = SimpleGraph(ones(Int8, n, n) - I )

    @time g3 = SimpleGraph(collect(iter))

    @time g4 = LightGraphs.SimpleGraphs.SimpleGraphFromIterator(iter)
    @time g5 = LightGraphs.SimpleGraphs.SimpleGraphFromIterator(collect(iter))
end
```
g1 uses the function `CompleteGraph`  and g2 creates a complete graph from an adjacency matrix. Both functions call `add_edge!` repeatedly.
g3 uses  the new constructor over a vector of edges.
g4 and g5 both use  `SimpleGraphFromIterator(iter)` . The difference is that g4 is created using a generator and g5 is created using a vector of edges.

For n = 10^4 I get the following results
```
g1:   19.551822 seconds (100.13 M allocations: 9.898 GiB, 13.41% gc time)
g2:   22.069623 seconds (100.13 M allocations: 10.934 GiB, 11.74% gc time)
g3:    3.666050 seconds (30.04 k allocations: 1021.695 MiB, 7.71% gc time)
g4:    6.664678 seconds (145.02 M allocations: 4.607 GiB, 25.56% gc time)
g5:    4.387188 seconds (140.04 k allocations: 2.699 GiB, 8.27% gc time)
```
Using a vector of edges is clearly the fastest method. This is, because in my implementation I first iterate over the vector in order to collect the degrees of the vertices and afterwards allocate the necessary space before adding the edges.

The iterator solution is  a bit slower. The reason for this is, that I don't iterate multiple times over the input and therefore have to relay on `push!` to allocate more memory if it is necessary. 
g5 is faster than g4 because when iterating over a vector I can get the iterator type at the start. Unfortunately this is not possible for generators at the moment, so I have to check the type of each element in the iterator.

Thoughts? Should I post the code here? Or should I maybe implement tests and create a pull request?

**Comment by somil55** [comment]
@sbromberger we can use `[kernel_1](https://github.com/JuliaGraphs/Graph500.jl/blob/master/src/kernel_1.jl)` function from Graph500 Package. It creates a graph given an edge list.

**Comment by sbromberger** [comment]
Closed via #880. Thanks!
